services/supervisor/src/transactions.rs

```diff
@@ -413,6 +447,150 @@ impl SupervisorInterface<ExecutionContext<'_>> for Supervisor {
+            // Store initial state of the request.
+            let state =
+                MigrationState::new(AsyncEventState::Pending, instance.data_version().clone());
+            schema.migration_states.put(&request, state.clone());
+            // Store the migration as pending. It will be removed in `before_transactions` hook
+            // once the migration will be completed (either successfully or unsuccessfully).
+            schema.pending_migrations.insert(request.clone());
+
+            // Finally, request core to start the migration.
+            // If migration initialization will fail now, it won't be a transaction execution error,
+            // since migration failure is one of possible outcomes of migration process. Instead of
+            // returning an error, we will just mark this migration as failed.
+            drop(schema);
+            let supervisor_extensions = context.supervisor_extensions();
+            let result = supervisor_extensions
+                .initiate_migration(request.new_artifact.clone(), request.service.as_ref());
+
+            // Check whether migration started successfully.
+            if let Err(error) = result {
+                // Migration failed even before start, softly mark it as failed.
+                let initiate_rollback = false;
+                return self.fail_migration(context, request, error, initiate_rollback);
+            }
+
+            // Migration started. Check if migration is fast-forward.
+            let instance = get_instance_by_name(&context, request.service.as_ref())
+                .expect("BUG: Instance disappeared");
+            if state.version != *instance.data_version() {
+                // Migration is fast-forward, complete it immediately.
+                // No agreement needed, since nodes which will behave differently will obtain
+                // different blockchain state hash and will be excluded from consensus.
+                log::trace!("Applied fast-forward migration with request {:?}", request);
+
+                let mut schema = SchemaImpl::new(context.service_data());
+
+                // Update the state of a migration.
+                let mut state = schema
+                    .migration_states
+                    .get(&request)
+                    .expect("BUG: Migration succeed, but does not have a stored state");
+                state.update(AsyncEventState::Succeed, instance.data_version().clone());
+                schema.migration_states.put(&request, state);
+
+                // Remove the migration from the list of pending.
+                schema.pending_migrations.remove(&request);
+            }
+        }
+        Ok(())
+    }
+
+    fn report_migration_result(
+        &self,
+        context: ExecutionContext<'_>,
+        result: MigrationResult,
+    ) -> Self::Output {
+        // Verifies that transaction author is validator.
+        let author = get_validator(&context)?;
+
+        result
```

Comment:
Nit: Not sure I understand why this check is necessary here. Logically, the artifact in the `result.request` was checked when a migration was initiated (which is checked later).

Reply:
It is not required indeed, but it allows us to spawn a more appropriate error IMHO.